### PR TITLE
r1407p-proposal-submission

### PR DESCRIPTION
**Pull request by r1407p**
---
A good PR should include the following parts:

- [x] Provide a clear title
- [x] Provide meaningful descriptions in the PR and the commit logs
- [x] Name your branch like ``<github-user-name>-hw#``, e.g., ``myname-hw2-attempt1``
- [x] Merge from your specific branch (***not master***) to the `hw*` branch (***not master***, again)
- [x] Exclude irrelevant files, such as binaries, text files, or dot files
- [x] Make sure tests/CI pass

### CODE REVIEW

File: project/r1407p/README.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
+NSD Term Project Proposal
```
Comment (yungyuc) [inline project/r1407p/README.md]: Why don't you use your project name as the proposal topic?

File: project/r1407p/README.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
+NSD Term Project Proposal
+---
+
+## Basic Information
+Project name : [PyMatrixOpt](https://github.com/r1407p/PyMatrixOpt)
+PyMatrixOpt is a Python matrix library with optimizations for improved performance. It provides various matrix operations and optimizations to enhance computational efficiency.
```
Comment (yungyuc) [inline project/r1407p/README.md]: One line is used as a paragraph and it is too long. It is hard to add comments at specific location for a long line. Please change to 79-char width.

File: project/r1407p/README.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+- Use ``CMake`` to build the binaries.
+- Use ``git`` for version control.
+- Use ``pytest`` for testing the performance and correctness.
+
+## Schedule
+- Week 1 (4/8): Research the matix optimization algorithm and implement the c++ version of the algorithm.
+- Week 2 (4/15): Implement the method in C++ class and write the function.
+- Week 3 (4/22): Finish the basic API function.
+- Week 4 (4/29): Add parallel computing to the algorithm.
+- Week 5 (5/6): Write unit test.
```
Comment (yungyuc) [inline project/r1407p/README.md]: Why don't you write unit tests in the first place?

**Review by yungyuc — CHANGES_REQUESTED** [review]
- [x] You missed an entry in `project/README.rst` as [requested](https://github.com/yungyuc/nsdhw_24sp/blob/567be93bc63a6fb1c2c55b8faf5825172daeaba5/project/README.rst?plain=1#L22)
- [x] Project name and proposal topic
- [x] 79-char line width
- [ ] Specify operations
- [x] Testing too late

It baffles to name your branch exactly the same as the upstream branch (project), but it's too late for you to change. Please avoid it in the future.

When you are ready for the next review, leave a comment, clearly write down something like "I am ready, please review again".

**Comment by yungyuc** [comment]
Don't resolve comments.  Your resolution makes it harder for me to keep track of them.

**Comment by yungyuc** [comment]
@r1407p Please update the proposal.

**Comment by r1407p** [comment]
I am ready, please review again
1. change to .rst file
2. set the project name
3. 79-char line width
4. change the specify operations in api
5. change the testing time

**Review by yungyuc — CHANGES_REQUESTED** [review]
- [x] You have not specify what operation to optimize.  Please update and highlight it using PR comment.
- [ ] Resolve branch conflict.

When you are ready for the next review, leave a comment, clearly write down something like "I am ready, please review again".  Do not forget the notification comment in any code review/PR.

**Comment by r1407p** [comment]
I am ready, please review again
the image is the highlight of operation to optimize
image

**Review by yungyuc — CHANGES_REQUESTED** [review]
Please highlight using PR comment and try again.  Reading the image, I still cannot know where to find the list in the text.  And do resolve conflict before requesting review again.

You did not seem to read my review comments carefully.  Please try better :)

File: project/r1407p/README.rst [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    Matrix matrix2 = Matrix::load_data(std::vector<std::vector<int>>{{7, 8}, {9, 10}, {11, 12}});
+    
+    matrix_naive = Matrix.matrix_multiply_naive(matrix1, matrix2);
+    matrix_naive_opt = Matrix.matrix_multiply_naive_opt(matrix1, matrix2); // PR: Highlight operation to optimize
+
+    matrix_strassen = Matrix.matrix_multiply_strassen(matrix1, matrix2);
+    matrix_strassen_opt = Matrix.matrix_multiply_strassen_opt(matrix1, matrix2); // PR: Highlight operation to optimize
+
+    matrix_coppersmith_winograd = Matrix.matrix_multiply_coppersmith_winograd(matrix1, matrix2);
+    matrix_coppersmith_winograd_opt = Matrix.matrix_multiply_coppersmith_winograd_opt(matrix1, matrix2); // PR: Highlight operation to optimize
```
Comment (r1407p) [inline project/r1407p/README.rst:105]: here are the operation to optimize.

Comment (r1407p) [reply project/r1407p/README.rst:105]: I am ready, please review again

Comment (yungyuc) [reply project/r1407p/README.rst:105]: Next time, please leave a **global comment** as I asked.  The inline comments are not obvious to find 🙂

**Review by yungyuc — APPROVED** [review]
It's now clear.  Thanks for the update, @r1407p

**Pull request merged**